api/handler.go

```diff
@@ -2,121 +2,66 @@ package api
 
 import (
 	"context"
-	"encoding/json"
 	"errors"
-	"fmt"
 	"net/http"
-	"net/http/httputil"
 	"path"
 	"strconv"
 	"time"
 
 	"github.com/golang/glog"
-	"github.com/google/uuid"
-	"github.com/julienschmidt/httprouter"
 	"github.com/livepeer/livepeer-data/health"
-	"github.com/livepeer/livepeer-data/health/reducers"
 	"github.com/livepeer/livepeer-data/pkg/data"
 	"github.com/livepeer/livepeer-data/pkg/jsse"
+	"github.com/nbio/hitch"
 )
 
 const (
 	sseRetryBackoff = 10 * time.Second
 	ssePingDelay    = 20 * time.Second
 	sseBufferSize   = 128
-	proxyLoopHeader = "X-Livepeer-Proxy"
-)
-
-type contextKey int
-
-const (
-	streamStatusKey contextKey = iota
 )
 
 type APIHandlerOptions struct {
 	APIRoot                       string
+	AuthURL                       string
 	RegionalHostFormat, OwnRegion string
 }
 
 type apiHandler struct {
-	opts        APIHandlerOptions
-	serverCtx   context.Context
-	core        *health.Core
-	regionProxy *httputil.ReverseProxy
+	opts      APIHandlerOptions
+	serverCtx context.Context
+	core      *health.Core
 }
 
 func NewHandler(serverCtx context.Context, opts APIHandlerOptions, healthcore *health.Core) http.Handler {
-	regionProxy := &httputil.ReverseProxy{
-		Director:      regionProxyDirector(opts.RegionalHostFormat),
-		FlushInterval: 100 * time.Millisecond,
-	}
-	handler := &apiHandler{opts, serverCtx, healthcore, regionProxy}
+	handler := &apiHandler{opts, serverCtx, healthcore}
 
-	router := httprouter.New()
-	router.HandlerFunc("GET", "/_healthz", handler.healthcheck)
+	router := hitch.New()
+	router.Use(cors)
+	router.HandleFunc("GET", "/_healthz", handler.healthcheck)
 
-	streamRoot := path.Join(opts.APIRoot, "/stream/:manifestId")
-	router.GET(streamRoot+"/health", handler.withRegionProxy(handler.getStreamHealth))
-	router.GET(streamRoot+"/events", handler.withRegionProxy(handler.subscribeEvents))
+	streamApiRoot := path.Join(opts.APIRoot, "/stream/:streamId")
+	middlewares := []hitch.Middleware{
+		streamStatus(healthcore, "streamId"),
+		regionProxy(opts.RegionalHostFormat, opts.OwnRegion),
+	}
+	if opts.AuthURL != "" {
+		middlewares = append(middlewares, authorization(opts.AuthURL))
```

Comment:
Does it matter at all that this is at the end of the middlewares list? Are we going to do a bunch of other streamStatus stuff before we realize that it's not allowed?

Reply:
So this was here to avoid authorizing a request twice when we do a region proxy. On the preceding middlewares we don't do much, only grab the stream ID from the URL then read the stream status from the healthcore, which reads from a `sync.Map` and should be most of the times lockless.

I have been thinking about moving the authorization earlier though, to possibly have more security here. Would be a little more inefficient due to the double authorization, but I think it shouldn't be much of an issue either anyway. 
e.g. to disallow snooping whether other users' streams exist, are active or even what region they are in (like a latency-based attack due to region proxy before a 403). 

WDYT?

Reply:
Kept it like this for simplicity in the code since both options have some drawbacks.